pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/GenericRow.java

```diff
@@ -120,6 +126,49 @@ public boolean hasNullValues() {
     return !_nullValueFields.isEmpty();
   }
 
+  /**
+   * @return a deep copy of the generic row
+   */
+  public GenericRow copy() {
+    GenericRow copy = new GenericRow();
+    copy.init(this);
+    for (Map.Entry<String, Object> entry : new HashSet<>(copy._fieldToValueMap.entrySet())) {
+      copy.putValue(entry.getKey(), copy(entry.getValue()));
+    }
+    return copy;
+  }
+
+  /**
+   * @return a deep copy of the object.
+   */
+  private Object copy(Object value) {
+    if (value == null) {
+      return null;
+    } else if (value instanceof Map) {
+      Map<String, Object> map = new HashMap<>();
+      for (String key : ((Map<String, Object>) value).keySet()) {
+        map.put(key, copy(((Map<String, Object>) value).get(key)));
+      }
```

Comment:
```suggestion
      Map<String, Object> map = new HashMap<>((Map<String, Object>) value);
      for (Map.Entry<String, Object> entry : map.entrySet()) {
        entry.setValue(copy(entry.getValue()));
      }
```

Reply:
this is less efficient?

Reply:
This will be more efficient because creating a map from another map is faster than adding each individual key-value pairs. After that we don't have extra map lookups